crypto/ec/ec_lib.c

```diff
@@ -274,6 +274,68 @@ int EC_METHOD_get_field_type(const EC_METHOD *meth)
 
 static int ec_precompute_mont_data(EC_GROUP *);
 
+/*-
+ * Try computing cofactor from the generator order (n) and field cardinality (q).
+ * This works for all curves of cryptographic interest.
+ *
+ * Hasse thm: q + 1 - 2*sqrt(q) <= n*h <= q + 1 + 2*sqrt(q)
+ * h_min = (q + 1 - 2*sqrt(q))/n
+ * h_max = (q + 1 + 2*sqrt(q))/n
+ * h_max - h_min = 4*sqrt(q)/n
+ * So if n > 4*sqrt(q) holds, there is only one possible value for h:
+ * h = \lfloor (h_min + h_max)/2 \rceil = \lfloor (q + 1)/n \rceil
+ *
+ * Otherwise, zero cofactor and return success.
+ */
+static int ec_guess_cofactor(EC_GROUP *group) {
+    int ret = 0;
+    BN_CTX *ctx = NULL;
+    BIGNUM *q = NULL;
+
+    if (group->field == NULL || group->order == NULL
+        || BN_is_zero(group->field) || BN_is_zero(group->order)
+        || BN_is_negative(group->field) || BN_is_negative(group->order)
+        || BN_num_bits(group->order) <= (BN_num_bits(group->field) + 1)/2 + 3) {
+        /*-
+         * If the cofactor is too large, we cannot guess it. Default to 0.
+         * The RHS of above is a strict overestimate of lg(4 * sqrt(q))
+         */
+        BN_zero(group->cofactor);
+        return 1;
+    }
+
+    if ((ctx = BN_CTX_new_ex(group->libctx)) == NULL)
```

Comment:
We're going to need a different version of this PR for all branches except master - `BN_CTX_new_ex` doesn't exist in those branches.